### DISCUSSION TITLE
`adjust-cell-height` calculation when moving between regular and hi-dpi displays

### DISCUSSION THREAD

**Discussion by crahan** [ghostty-org/ghostty#3608] [Q&A]
I've currently set `adjust-cell-height` to `5%` in my Ghostty configuration and I noticed that when I move the Ghostty window between a regular resolution external monitor and the Macbook Pro's Retina display, Ghostty renders the same content slightly less high. 

The following screenshots hopefully explain the issue a little better. The back most window is iTerm, then macOS Terminal.app, then Ghostty. Here's the 3 applications on the regular external monitor. Both Ghostty and iTerm have a 5% line-height increase from the default. All 3 applications render the `ls -l` output at the same height.

![Screenshot 2024-12-27 at 23 53 44](https://github.com/user-attachments/assets/7aef6e51-ef1a-446f-8d8d-dd10c15c7dba)

When I move the 3 windows to the Retina display, Terminal.app and iTerm still render things at the same height, but Ghostty now renders the `ls -l` output slightly less high compared to the other 2 apps.

<img width="852" alt="Screenshot 2024-12-27 at 23 54 37" src="https://github.com/user-attachments/assets/913ee3b2-f74b-4324-a142-8aee3df59cc7" />

I know absolutely nothing about font rendering, so I can only guess at what's going on. What I did notice is that when I change `adjust-cell-height` to `10%`, everything aligns on the Retina display. This made me wonder if the initial cell height is calculated when the Ghostty window is rendered on the external monitor, but isn't updated when the window is moved to the Retina display, resulting in the `adjust-cell-height` value being half of what it should be for the Retina display?

**Comment by crahan** [comment]
These additional screenshots show the Ghostty window rendered on the external (regular) monitor and the hi-dpi Retina display. One image is 72dpi at 746px height, while the other is 144dpi at 1492px height. Dragging the window to the Retina display increases the gap between the bottom of the content and the bottom of the window.

![Screenshot 2024-12-28 at 00 27 28](https://github.com/user-attachments/assets/e1f187b2-ce8c-4a0c-b9a3-e98072b18cd1)

<img width="746" alt="Screenshot 2024-12-28 at 00 27 33" src="https://github.com/user-attachments/assets/a7b8e423-f71b-4622-86f9-c2716f5cfeae" />

As I flip between these two images, I'm honestly unsure if the height difference is because of how `adjust-cell-height` is handled or if the font itself is rendered at a different height on each display.

**Comment by 00-kat** [comment]
I can't tell if I can reproduce this with `adjust-cell-height`/`adjust-cell-width`, but I do seem to have differences between a hi-dpi display and a regular one with `adjust-cursor-thickness` on Linux.  Not sure if that's relevant to the same issue, though; I'm only mentioning it because it's another `adjust-` option.

**Comment by zeevenn** [comment]
Most obvious in vim or tmux statusline. There's some unnecessary stuff at the bottom.

<img width="440" height="32" alt="image" src="https://github.com/user-attachments/assets/7f235f62-d9fa-4aa4-b105-5a59d3b708da" />

**Reply by danielwe** [reply]
Terminals render characters in a fixed grid where the cell size is calculated from the font. That's why you get gaps like this around the edges, unless the window dimensions happen to be exact integer multiples of the cell dimensions. The way to get rid of this is to resize the window.

You can also take a look at the `window-padding-color = extend` config option (https://ghostty.org/docs/config/reference#window-padding-color), but the chevron shapes in your statusline prevent the colors from being extended downward in this case.

This is unrelated to the DPI-related rounding differences discussed here (which in any case work differently now than when this discussion was active).

**Reply by zeevenn** [reply]
Thanks for reply.

If I want to increase the line spacing so that the lines look more relaxed, how should I approach this?

My config is here:

```
font-size = 14
theme = Catppuccin Mocha

macos-titlebar-style = transparent

font-family = "JetBrainsMono Nerd Font"
# CJK fallback
font-family = "Noto Sans CJK SC"

font-codepoint-map = "U+2700-U+27BF=Zapf Dingbats"

confirm-close-surface = false
quit-after-last-window-closed = true

window-padding-balance = true
window-padding-y = 0
window-padding-x = 0
window-padding-color = extend

window-save-state = always

fullscreen = true

adjust-cell-height = 7

font-feature = -calt
```

**Reply by danielwe** [reply]
Line spacing is controlled by `adjust-cell-height`. Currently, you're setting it to an integer, which is interpreted as a number of pixels and therefore gives different results on monitors with different DPI. I'd recommend using a percentage instead. Start with `adjust-cell-height = +10%` and adjust up/down from there.

**Reply by zeevenn** [reply]
Seems it also can't fix this, even if I set it to `0` or `any percent`. In different screen it will always have some gap in the bottom.

**Reply by danielwe** [reply]
I see, what you're saying is that the gap will be different on screens with different DPI. Yes, this is expected and unavoidable. It comes down to the pixel rounding of the cell dimensions.

In some cases, it's may that the difference is exacerbated due to the `adjust-cell-height` adjustment factor being applied _after_ the rounding step, but even if we changed this, there would generally be _some_ difference. See the comment below (https://github.com/ghostty-org/ghostty/discussions/3608#discussioncomment-15378912).

If you think this is something that truly ought to be improved, please open a new discussion in the Issue Triage category and provide some more detail and screenshots to explain why you're unhappy with the current state.

**Reply by danielwe** [reply]
@zeevenn It just dawned on me you may be interested in the config option `window-step-resize = true` (https://ghostty.org/docs/config/reference#window-step-resize). It should eliminate the gap along the edges by resizing the window as needed to exactly match the grid. Currently macOS only.

**Reply by zeevenn** [reply]
I have tried but seems can't fix this question.

**Comment by danielwe** [comment]
Since this discussion was revived, let me briefly address the points in the OP to provide some closure:

* The cell dimensions of the character grid are extracted from the font and then rounded to an integer number of pixels. The rounding is needed such that every cell is equivalent to every other cell for rasterization and rendering purposes.
* On a HiDPI monitor, the pixel grid we round to is about twice as fine-grained as on a low DPI monitor. This is why the relative cell size and aspect ratio may differ slightly between DPIs. On the HiDPI monitor you can get closer to the font's ground truth.
* `adjust-cell-{width,height}` apply to the dimensions _after_ rounding. The result is then rounded again. This keeps the implementation simple, but it introduces another DPI-dependent nonlinearity affecting the final cell size. The result may be that DPI-related differences behave slightly less intuitively than you might expect in some cases. Not sure if that's the key issue here?
* These calculations don't work exactly the same now as they did when this post was active. See #9432 and #9648 for the most recent changes.

With that, I'll close this discussion. Please open a new QA or Issue Triage discussion for specific concerns with one or more of these points pertaining to Ghostty v1.2.3 or newer.
